gzochi-server/doc/gzochi-load.1
.TH GZOCHI-LOAD 1 "May 4, 2015"
.SH NAME
gzochi-load \- Load exported gzochi databases
.SH SYNOPSIS
.B gzochi-load
[-c <conf>] [-e <engine>] [-f] <app_name or data_dir>:<db>
.br
.B gzochi-load
[-h | -v]
.SH DESCRIPTION
The gzochi-load utility reads rows of data from standard input and loads them 
into the specified gzochi game application database. 
.PP
The input to gzochi-load must be in the input format specified by the 
\fBgzochi-dump\fR utility. This format is very similar to the format used by the
dump and load tools for Berkeley DB, though gzochi-load can read the output of 
the other gzochid tools no matter what storage engine the server is built 
against.
.PP
The argument to gzochi-load is the name of a gzochi game application database 
(i.e., \fBmeta\fR, \fBoids\fR, or \fBnames\fR) into which the data should be
loaded, qualified by an application name or the path to a directory containing 
an application's databases. If gzochi-load detects that the first part of the
argument is not a directory, it assumes it must be an application name, and will
consult the server's gzochid.conf file to obtain the root of the server's
application data directory tree, which it uses to resolve the full path to the
application's database files. Whether specified as a path or an application 
name, the target directory need not already exist; gzochi-load will create it. 
The command:

.IP
gzochi-load /var/gzochid/data/mygame2:meta < meta.dump
.PP

will load the contents of the file \(dqmeta.dump\(dq into the \(dqmeta\(dq 
database in the specified directory. All three databases are necessary to fully 
reconstitute the state of a gzochi game application. For example, the commands:

.IP
gzochi-load mygame2:meta < meta.dump
.br
gzochi-load mygame2:names < names.dump
.br
gzochi-load mygame2:oids < oids.dump
.PP

...will load each of the required databases from its constituent dump file. Note
that by default gzochi-load will refuse to load data into a database that 
already exists.

.SH OPTIONS
.IP \fB\-c,\ \-\-config\fR
Specify an alternate path to the gzochid.conf file.
.IP \fB\-e,\ \-\-engine\fR
Specify the name of the storage engine to use to load data into the target 
database. The format of this argument is the same as for the 
\(dqstorage.engine.dir\(dq setting in gzochid.conf: the basename of the 
module file, minus the extension. (E.g., bdb.)
.IP \fB\-f,\ \-\-force\fR
Disable the safeguards that prevent pre-existing database files from being
written to. If this flag is specified, gzochi-load will add the contents of
the dump file to the target database without truncating it. Use with care!
.IP \fB\-h,\ \-\-help\fR
Write usage information to standard error, and exit.
.IP \fB\-v,\ \-\-version\fR
Write version information to standard error, and exit.

.SH ENVIRONMENT
.TP
.B GZOCHID_CONF_LOCATION
If \fB$GZOCHID_CONF_LOCATION\fR is set before \fBgzochi-load\fR is started, its
value is used as the full path to the server configuration file. This variable 
will be overriden by a command-line argument giving the location of the file, if
present.
.TP
.B GZOCHID_STORAGE_ENGINE_DIR
If \fB$GZOCHID_STORAGE_ENGINE_DIR\fR is set before \fBgzochi-load\fR is started,
its value is used as the base path for resolving storage engine modules. This
value takes precedence over the \(dqstorage.engine.dir\(dq setting in the server
configuration file.

.SH AUTHOR
Julian Graham (joolean@undecidable.net)


NEWS
gzochi NEWS -- history of user-visible changes
Copyright (C) 2015 Julian Graham
See the end of copying conditions.

Send comments, questions, and bug reports to gzochi-devel@nongnu.org

Changes in 0.8 (since 0.7)

* Changes to the distribution

** Removed deprecated synchronous client (libgzochi)

libgzochi, the synchronous reference client implementation included with earlier
versions of the gzochi framework, has been removed from the distribution. The
client was deprecated in an earlier release, since the concurrency strategy
suggested by its API was undesirable. Application developers are encouraged to
use libgzochi-glib instead.

** Removed client support static library (libgzochi-client-common)

libgzochi-client-common, the static library that facilitated sharing of code
between libgzochi and libgzochi-glib, has been merged with libgzochi-glib and no
longer exists as an independent build target within the gzochi framework source
distribution.

** New example: External transactions

Client and server code for a new example application has been added to the 
source distribution under the "examples/" tree. This example demonstrates 
techniques for synchronizing a transaction executed within the gzochid container
with an external transaction-aware process, such as a relational database.


Changes in 0.7 (since 0.6)

* Changes to the distribution

** New reference client library implementation: (gzochi client)

Game developers who want to keep their entire application stack in Guile Scheme
(C server container notwithstanding) are in luck! The new (gzochi client)
library is a Scheme reference client implementation with an execution model that
can be integrated into a main loop or used stand-alone. The source distribution
for (gzochi client) is under the "gzochi-guile-client/" tree.


Changes in 0.6 (since 0.5)

* Compatibility

** Support for additional platforms

gzochi now builds on Mac OS X and (likely) other BSD variants.


Changes in 0.3 (since 0.2)

* Changes to the distribution

** New client support static library: libgzochi-client-common

Some of the trickier aspects of client communication with a gzochi server
instance, such as maintaining a buffer for partial message delivery, have been
factored out of libgzochi and into a static library, with a source distribution
under the "gzochi-client-common/" tree, that can be built into other client 
implementations. 

** New reference client library implementation: libgzochi-glib

libgzochi-client-common already has a consumer in the distribution:
libgzochi-glib, a new reference client implementation that integrates easily 
with the "main loop" functionality in the popular GNU GLib library to support
the development of single-threaded client applications. The source distribution
for libgzochi-glib is under the "gzochi-glib-client/" tree.

** Removed deprecated debugging client

mictlan, the primitive "debugging client" included with earlier versions of the
gzochi framework, has been removed from the distribution. Its functionality has
been superceded by the debugging capabilities of gzochid itself, which can be 
accessed using a telnet client.


Changes in 0.2 (since 0.1)

* Changes to the distribution

** New example: abermud

Client and server code for a new example game -- based on Alan Cox's 
AberMUD -- has been added to the source distribution under the "examples/"
tree. This example demonstrates techniques for maintaining a persistent world
in which state is shared between multiple players.


Copyright information:

Copyright (C) 2015 Julian Graham

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.


Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


gzochi-server/src/gzochi-load.c
/* gzochi-load.c: Utility for importing data from game application databases
 * Copyright (C) 2015 Julian Graham
 *
 * gzochi is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>
#include <errno.h>
#include <getopt.h>
#include <glib.h>
#include <locale.h>
#include <libintl.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "gzochid-storage.h"
#include "toollib.h"

#define _(String) gettext (String)

struct load_context
{
  gzochid_storage_engine *engine;
  gzochid_storage_context *storage_context;
  gzochid_storage_store *store;
};

struct datum
{
  char *data;
  size_t data_len;
};

static struct load_context *
setup_load_context (gzochid_storage_engine *engine, char *data_dir, char *db, 
		    gboolean force)
{
  unsigned int flags = GZOCHID_STORAGE_CREATE | GZOCHID_STORAGE_EXCL;
  struct load_context *context = calloc (1, sizeof (struct load_context));
  char *path = g_strconcat (data_dir, "/", db, NULL);
 
  context->engine = engine;
  context->storage_context = engine->interface->initialize (data_dir);

  if (context->storage_context == NULL)
    {
      g_critical ("Failed to initialize store in %s", data_dir);
      exit (EXIT_FAILURE);
    }

  if (force)
    flags ^= GZOCHID_STORAGE_EXCL;

  context->store = engine->interface->open 
    (context->storage_context, path, flags);

  if (context->store == NULL)
    {
      g_critical ("Failed to open store in %s", data_dir);
      exit (EXIT_FAILURE);
    }

  free (path);
  
  return context;
}

static void
load_data_header (char *line, size_t line_len, int line_num)
{
}

static char *
pack_line (char *line, size_t len, int line_num)
{
  size_t i = 0;
  char *packed_line = calloc (len / 2, sizeof (char));

  for (; i < len; i += 2)
    if (sscanf (line + i, "%2hhx", &packed_line[i / 2]) != 1)
      {
	g_critical ("Malformed line at row %d.", line_num);
	exit (EXIT_FAILURE);
      }

  return packed_line;
}

static struct datum
read_line (char *line, size_t line_len, int line_num)
{
  struct datum data;

  line_len -= 2;
  
  if (line_len <= 0 || line_len % 2 != 0)
    {
      g_critical ("Malformed line at row %d.", line_num);
      exit (EXIT_FAILURE);
    }
  
  data.data = pack_line (line + 1, line_len, line_num);
  data.data_len = line_len / 2;
      
  return data;
}

static void
cleanup_load_context (struct load_context *context)
{
  context->engine->interface->close_store (context->store);
  context->engine->interface->close_context (context->storage_context);
  
  free (context);
}

static void
load_data (gzochid_storage_engine *engine, char *data_dir, char *db, 
	   gboolean force)
{
  int line_num = 1;
  struct datum line = { NULL, 0 }, key, value;
  gboolean reading_key = TRUE;
  struct load_context *context = 
    setup_load_context (engine, data_dir, db, force);

  while ((line.data_len = getline (&line.data, &line.data_len, stdin)) != -1 
	 && strcmp (line.data, "HEADER=END\n") != 0)
    {
      load_data_header (line.data, line.data_len, line_num++);
      
      free (line.data);
      line.data = NULL;
      line.data_len = 0;
    }

  if (line.data_len == -1)
    {
      g_critical ("While reading header data: %s", strerror (errno));
      exit (EXIT_FAILURE);
    }

  line_num++;

  while ((line.data_len = getline (&line.data, &line.data_len, stdin)) != -1 
	 && strcmp (line.data, "DATA=END\n") != 0)
    {
      if (reading_key)
	{
	  key = read_line (line.data, line.data_len, line_num++);
	  reading_key = FALSE;
	}
      else 
	{
	  value = read_line (line.data, line.data_len, line_num++);
	  context->engine->interface->put 
	    (context->store, key.data, key.data_len, value.data, 
	     value.data_len);

	  free (key.data);
	  free (value.data);

	  reading_key = TRUE;
	}

      free (line.data);
      line.data = NULL;
      line.data_len = 0;
    }

  cleanup_load_context (context);
}

static const struct option longopts[] =
  {
    { "config", required_argument, NULL, 'c' },
    { "engine", required_argument, NULL, 'e' },
    { "force", no_argument, NULL, 'f' },
    { "help", no_argument, NULL, 'h' },
    { "version", no_argument, NULL, 'v' },
    { NULL, 0, NULL, 0 }
  };

static void
print_version (void)
{
  fprintf (stderr, "gzochi-load (gzochi) %s\n", VERSION);

  fputs ("", stderr);
  fprintf (stderr, _("\
Copyright (C) %s Julian Graham\n\
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n\
This is free software: you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.\n"),
          "2015");
}

static void
print_help (const char *program_name)
{
  fprintf (stderr, _("\
Usage: %s [-c <CONF>] [-e <ENGINE>] [-f] <APP_NAME or DATA_DIR>:<DB>\n\
       %s [-h | -v]\n"), program_name, program_name);

  fputs ("", stderr);
  fputs (_("\
  -c, --config        full path to gzochid.conf\n\
  -e, --engine        the name of the storage engine in use, as per\n\
                      gzochid.conf (e.g., bdb)\n\
  -f, --force         force an import, even when the target already exists\n\
  -h, --help          display this help and exit\n\
  -v, --version       display version information and exit\n"), stderr);

  fputs ("", stderr);
  fprintf (stderr, _("\
Report bugs to: %s\n"), PACKAGE_BUGREPORT);
#ifdef PACKAGE_PACKAGER_BUG_REPORTS
  fprintf (stderr, _("Report %s bugs to: %s\n"), PACKAGE_PACKAGER,
          PACKAGE_PACKAGER_BUG_REPORTS);
#endif /* PACKAGE_PACKAGER_BUG_REPORTS */

#ifdef PACKAGE_URL
  fprintf (stderr, _("%s home page: <%s>\n"), PACKAGE_NAME, PACKAGE_URL);
#else
  fprintf (stderr, _("%s home page: <http://www.nongnu.org/%s/>\n"),
	   PACKAGE_NAME, PACKAGE);
#endif /* PACKAGE_URL */
}

int 
main (int argc, char *argv[])
{
  const char *program_name = argv[0];
  char *gzochid_conf_path = NULL;
  char *storage_engine_name = NULL;
  gboolean force = FALSE;
  int optc = 0;
  
  setlocale (LC_ALL, "");
  
  while ((optc = getopt_long (argc, argv, "+ce:fhv", longopts, NULL)) != -1)
    switch (optc)
      {
      case 'c':
	gzochid_conf_path = strdup (optarg);
	break;
      case 'e':
	storage_engine_name = strdup (optarg);
	break;       
      case 'f':
	force = TRUE;
	break;
	
      case 'v':
	print_version ();
	exit (EXIT_SUCCESS);
	break;
      case 'h':
	print_help (program_name);
	exit (EXIT_SUCCESS);
	break;

      default:
	print_help (program_name);
	exit (EXIT_FAILURE);
      }

  if (optind != argc - 1)
    {
      print_help (program_name);
      exit (EXIT_FAILURE);
    }
  else
    {
      GHashTable *gzochid_conf = 
	gzochid_tool_load_game_config (gzochid_conf_path);

      char **targets = gzochid_tool_parse_targets (argv[optind]);
      char *data_dir = gzochid_tool_probe_data_dir 
	(gzochid_conf, targets[0], TRUE);
      char *db = targets[1];

      gzochid_storage_engine *engine = 
	gzochid_tool_probe_storage_engine (gzochid_conf, storage_engine_name); 

      if (db == NULL)
	{
	  print_help (program_name);
	  exit (EXIT_FAILURE);
	}
      else load_data (engine, data_dir, db, force);
      g_strfreev (targets);
    }

  return 0;
}
